### PR TITLE
Provide additional workspace API to add/remove workspace folders (for #35407)

### PR DESCRIPTION
**Pull request by bpasero**
@jrieken here is the suggested API for adding/removing workspace folders, feedback welcome.

**Note** that I am still planning to tackle 2 outstanding issues independent of this PR to keep the discussion on the API shape only:
* when removing a workspace folder from a single-folder workspace, we should probably turn the workspace into an empty one (same as File > Close Folder)
* when adding a workspace folder while you are in empty/single-folder workspace we should transition into a workspace (same as we do today via File > Add Folder To Workspace...).

### CODE REVIEW

File: src/vs/vscode.d.ts [hunk 1]
```diff
@@ -5199,6 +5199,32 @@ declare module 'vscode' {
...
+		 * Adds workspace folders to the currently opened workspace. Will ignore workspace
+		 * folders that are already part of the workspace.
+		 *
+		 * Note: if this workspace had no folder opened, all extensions will be restarted
+		 * so that the (deprecated) `rootPath` property is updated to point to the first workspace
+		 * folder.
+		 *
+		 * @param folders a list of workspace folders to add.
+		 */
+		export function addWorkspaceFolders(folders: { uri: Uri, name?: string }[]): Thenable<void>;
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: Is adding many folders at once really a use-case? I'd expect this to be called for single folders in 99% of the cases. My proposal would be `function addWorkspaceFolder(uri: Uri, name?: string)`

File: src/vs/vscode.d.ts [hunk 2]
```diff
@@ -5199,6 +5199,32 @@ declare module 'vscode' {
...
+		 *
+		 * This method will be a no-op when called while not having a workspace opened.
+		 *
+		 * Note: if the first workspace folder is removed, all extensions will be restarted
+		 * so that the (deprecated) `rootPath` property is updated to point to the first workspace
+		 * folder.
+		 *
+		 * @param folders a list of [workspace folders](#WorkspaceFolder) to remove.
+		 */
+		export function removeWorkspaceFolders(folders: WorkspaceFolder[]): Thenable<void>;
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: Same, N vs 1

File: src/vs/workbench/api/node/extHostWorkspace.ts [hunk 3]
```diff
@@ -78,6 +78,17 @@ export class ExtHostWorkspace implements ExtHostWorkspaceShape {
 		}
 	}
 
+	addWorkspaceFolders(folders: { uri: URI, name?: string }[]): Thenable<void> {
+		return this._proxy.$addFolders(folders);
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostWorkspace.ts]: Not yet decided if those changes should be reflected sync... We do that in some place, e.g the editor or Memento#update, where we assume optimistically that the change is good, then send to the main side which might cause another change... Needs some thinking.

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/vscode.d.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/vscode.d.ts]: Fixed via https://github.com/Microsoft/vscode/pull/36820/commits/158d1520d6823b4584f686f564b2a35ace6d4c55

File: src/vs/vscode.d.ts (hunk 2 above)
Comment (bpasero) [reply src/vs/vscode.d.ts]: Fixed via https://github.com/Microsoft/vscode/pull/36820/commits/158d1520d6823b4584f686f564b2a35ace6d4c55

File: src/vs/workbench/api/node/extHostWorkspace.ts (hunk 3 above)
Comment (bpasero) [reply src/vs/workbench/api/node/extHostWorkspace.ts]: @jrieken well in some of the cases this action will result in the extension host restarting, so in that case it does not really matter. For the other case of an already opened workspace, I think even with the current promise chaining there is a chance that the `onDidChangeWorkspaceFolders` has not yet fired when the method returns.

What is your suggestion?

File: src/vs/workbench/api/node/extHostWorkspace.ts [hunk 4]
```diff
@@ -78,6 +78,18 @@ export class ExtHostWorkspace implements ExtHostWorkspaceShape {
 		}
 	}
 
+	addWorkspaceFolder(uri: URI, name?: string): Thenable<void> {
+		return this._proxy.$addFolder(uri, name);
+	}
+
+	removeWorkspaceFolder(folder: vscode.WorkspaceFolder): Thenable<void> {
+		if (this.getWorkspaceFolders().indexOf(folder) === -1) {
+			return Promise.resolve(undefined);
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostWorkspace.ts]: Maybe make this return a boolean to indicate success/failure. Might also be useful when having a user-confirmation in between. Then `false` could also mean blocked by the user

Comment (bpasero) [reply src/vs/workbench/api/node/extHostWorkspace.ts]: @jrieken yeah added via https://github.com/Microsoft/vscode/pull/36820/commits/1dc821b614609990436008b2aff5fd262a343b09

**Comment by bpasero** [comment]
@jrieken the corner cases are now covered such as the API call will work in any workspace state (empty workspace, single-folder or multi-root workspace).

There is one remaining issue (related to your comment about running this sync or not): When the extension host needs to be restarted as a matter of the call (this can happen when adding a root folder while either no workspace is opened or when a single-folder is opened), the extension host goes down _while_ executing the request and the result is an ugly error message:

image

The only way how to prevent this message is to run the workspace edit operation in a timeout and return to the extension directly without waiting for the operation to be finished.

Undecided what to do here, I feel that executing the command sync and waiting for it is correct and we should maybe add code to not show this error message in that case.

**Comment by jrieken** [comment]
> workspace edit operation in a timeout and return to the extension directly without executing and waiting for the operation to be finished.

Not nice... In the does-not-need-to-restart case we should ensure that the workspace change event happened in-between making the call and resolving the promise. Timeout-trick will likely break that. What of the call is causing the cancel-error? Is this us calling back to the extension host?

File: src/vs/vscode.d.ts [hunk 5]
```diff
@@ -5211,6 +5211,34 @@ declare module 'vscode' {
 		export const onDidChangeWorkspaceFolders: Event<WorkspaceFoldersChangeEvent>;
 
 		/**
+		 * Adds a workspac folder to the currently opened workspace.
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: e

Comment (bpasero) [reply src/vs/vscode.d.ts]: via https://github.com/Microsoft/vscode/pull/36820/commits/95eccd67c91c503e2ee00ba308d001806108c47e

**Comment by bpasero** [comment]
@jrieken the `onDidChangeWorkspaceFolders` is firing before the method returns with the current code. Agree that a timeout is just ugly. 

The "Canceled" error seems to originate from [`rpcProtocol#dispose()`](https://github.com/Microsoft/vscode/blob/ben/35407/src/vs/workbench/services/extensions/node/rpcProtocol.ts#L41).

image

**Comment by HutzlerGregg** [comment]
Give as we doing and I'll.

**Comment by jrieken** [comment]
> The "Canceled" error seems to originate from rpcProtocol#dispose().

Don't we handle those standard-cancellation-errors everywhere? And by "handle" I meaning "ignore"?

**Comment by bpasero** [comment]
@jrieken yeah looks like the commands list was not checking for that, I added it via https://github.com/Microsoft/vscode/pull/36820/commits/48c451129abce37e221c2e7d657a715b2f9d395d

Can we merge this in so that we have it for October or do you want more changes? I was still planning to look into the confirmation dialog thing but can do that outside of this PR.

**Comment by bpasero** [comment]
Added the confirm dialog via https://github.com/Microsoft/vscode/pull/36820/commits/42e6cd58461655e9d88d062f982423a50f4cbcc5 with a new setting `workbench.confirmChangesToWorkspaceFromExtensions`.

image

**Pull request merged**